escrow/combined_storage.go

```diff
@@ -0,0 +1,177 @@
+	if ok && err == nil {
+		return
+	}
+	if err != nil {
+		return nil, false, err
+	}
+	log.Info("Channel key is not found in storage")
+
+	state, ok, err = storage.getChannelStateFromBlockchain(key.ID)
+	if !ok || err != nil {
+		return
+	}
+	log = log.WithField("state", state)
+	log.Info("Channel found in blockchain")
+
+	// TODO: see comment at the beginning of the method
+	if state.Nonce.Cmp(key.Nonce) != 0 {
+		log.Warn("Channel nonce is not equal to expected")
+		return nil, false, fmt.Errorf("Channel nonce: %v is not equal to expected: %v", state.Nonce, key.Nonce)
+	}
+
+	ok, err = storage.CompareAndSwap(key, nil, state)
+	if err != nil {
+		return
+	}
+	if !ok {
+		log.Warn("Key is already present in the storage")
+		return nil, false, err
+	}
+	log.WithField("state", state).Info("Channel saved in storage")
+
+	return
+}
+
+var zeroAddress = common.Address{}
+
+func (storage *combinedStorage) getChannelStateFromBlockchain(id *big.Int) (state *PaymentChannelData, ok bool, err error) {
+	log := log.WithField("id", id)
+
+	channel, err := storage.mpe.Channels(nil, id)
+	if err != nil {
+		log.WithError(err).Warn("Error while looking up for channel id in blockchain")
+		return nil, false, err
+	}
+	if channel.Sender == zeroAddress {
+		log.Warn("Unable to find channel id in blockchain")
+		return nil, false, nil
+	}
+	log = log.WithField("channel", channel)
+	log.Debug("Channel found in blockchain")
+
+	configGroupId := config.GetBigInt(config.ReplicaGroupIDKey)
+	if channel.ReplicaId.Cmp(configGroupId) != 0 {
+		log.WithField("configGroupId", configGroupId).Warn("Channel received belongs to another group of replicas")
+		return nil, false, fmt.Errorf("Channel received belongs to another group of replicas, current group: %v, channel group: %v", configGroupId, channel.ReplicaId)
+	}
+
+	return &PaymentChannelData{
+		Nonce:            channel.Nonce,
+		State:            Open,
+		Sender:           channel.Sender,
+		Recipient:        channel.Recipient,
+		GroupId:          channel.ReplicaId,
+		FullAmount:       channel.Value,
+		Expiration:       time.Unix(channel.Expiration.Int64(), 0),
+		AuthorizedAmount: big.NewInt(0),
+		Signature:        nil,
+	}, true, nil
+}
+
+func (storage *combinedStorage) Put(key *PaymentChannelKey, state *PaymentChannelData) (err error) {
+	return storage.delegate.Put(key, state)
+}
+
+func (storage *combinedStorage) CompareAndSwap(key *PaymentChannelKey, prevState *PaymentChannelData, newState *PaymentChannelData) (ok bool, err error) {
+	return storage.delegate.CompareAndSwap(key, prevState, newState)
+}
+
+type memoryStorageKey string
+
+type memoryStorage struct {
+	data map[memoryStorageKey]*PaymentChannelData
+}
+
+func NewMemStorage() (storage PaymentChannelStorage) {
+	return &memoryStorage{
+		data: make(map[memoryStorageKey]*PaymentChannelData),
+	}
+}
+
+func getMemoryStorageKey(key *PaymentChannelKey) memoryStorageKey {
+	return memoryStorageKey(fmt.Sprintf("%v", key))
```

Comment:
GOB is also not tolerant to the changing of order of the fields. We can introduce PaymentChannelKey method Bytes to return binary key representation. And use determined order of fields for compatibility.

Reply:
I have found that actually order of fields here doesn't matter because I have implemented PaymentChannelKey.String(): https://github.com/singnet/snet-daemon/blob/9e4c76594531713ff54fa8b468cc593892d1dfde/escrow/escrow.go#L94